Implement driver for power management

rohanpurohit opened the pull request:
This PR includes:
- added a low driver for LTC4015 with following functions:
  - is_charger_enabled
  - read_battery_voltage
  - read_input_voltage
  - read_system_voltage
  - read_battery_current
  - read_input_current
- Added a high level driver that currently just checks if the charger is enabled and triggers an alarm
- moved AlarmsManager from BreathingCircuit to Application
- added a overloaded function for actiavte_alarm with no alarm_limits
- added a new LogEventCode - charger_disconnected
- made changes to add BatteryPower messagetype

rohanpurohit committed:
initial implementation of power management low level driver

rohanpurohit committed:
HAL: I2C: add methods to write to a specific register

rohanpurohit committed:
clean-up the driver for prototype phase

rohanpurohit committed:
LTC4015: implement charger enabled method

rohanpurohit committed:
Merge 'develop' into feature/firmware-driver-power-management

rohanpurohit committed:
add battery power protobuf message and update alarm manager

rohanpurohit committed:
firmware: generate an alarm if charger is disconnected

rohanpurohit committed:
firmware: add schedule entry for battery power message type

rohanpurohit committed:
backend: add battery power state segment to frontend synchroniser schedule

rohanpurohit committed:
generate nanopb

rohanpurohit committed:
HALI2Cdevice: remove HAL_IO_Mem_write abstraction

ethanjli commented:
From an overview-level code review, this looks good. We discussed moving alarms-related functionality out of the high-level Sensor class for LTC4015, and changing the function interface and name to `read_charging_status` and renaming the `BatteryPower` message field from `charging_status` to `charging`. The next step is to make any necessary fixes once hardware gets tested, then I'll do another quick review and we can merge.

rohanpurohit committed:
STM32: Fix HAL_I2C_Mem_Read for LTC4015

* doesn't work if device address is left shifted by 1

rohanpurohit committed:
move alarm related code out of high level driver

rohanpurohit committed:
update BatteryPower message type to Battery

rohanpurohit committed:
Power: add power related alarm code into its own module

rohanpurohit committed:
apply formatting auto fixes

rohanpurohit committed:
fix battery msg type declarations

rohanpurohit committed:
fix clang-tidy errors

rohanpurohit commented:
Changes:
- Found in testing that while doing read from HAL_I2C_Mem_Read, failed when the device address was left-shifted by 1, thus the abstraction was changed to fix that.
- Renamed BatteryPower MessageType to Battery
- Renamed charging_status field of battery to charging
- moved alarm related code from the high-level sensor driver to power module 
- update read_charging_status method in low-level driver

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-driver-power-management

rohanpurohit committed:
proto: rename Battery to PowerManagement

rohanpurohit committed:
fix frontend and backend errors

rohanpurohit committed:
update PowerManagement type code

rohanpurohit committed:
backend: add power management service to simulator

rohanpurohit committed:
frontend: fix type code in frontend for power management

rohanpurohit committed:
firmware: fix trailing whitespaces

rohanpurohit committed:
frontend: Add eventType for charger disconnected

rohanpurohit committed:
update power management simulator

rohanpurohit committed:
add power management simulator to firmware

rohanpurohit committed:
ltc4015: update high level driver

rohanpurohit committed:
firmware: update power management simulator

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-driver-power-management

rohanpurohit committed:
firmware: deactive alarms when not ventilating

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-driver-power-management

rohanpurohit committed:
fix backend lint error

rohanpurohit committed:
firmware: decouple alarm codes from alarms.h

rohanpurohit committed:
firmware: revert IsDeviceReady implementation

rohanpurohit committed:
firmware:Sensor.h: remove wait_measurement action and cleanup

rohanpurohit committed:
backend: cleanup and improve backend power simulator

rohanpurohit committed:
update power simulator in firmware

rohanpurohit committed:
firmware: fix clang-tidy errors

rohanpurohit committed:
update power simulator

rohanpurohit committed:
Revert "firmware:Sensor.h: remove wait_measurement action and cleanup"

This reverts commit d681c9e006b47a0070ce64bb3862b333e2900986.

ethanjli commented:
Results from testing on commit 875b8ed, with `ltc4015_sensor` removed from the list of firmware initializables:
1. the firmware behaves as expected: battery level starts from 0% and increases at a reasonable rate to 100% but then stays there, and a "Charger is disconnected" alarm is active once it's at 100%.
2. When I remove the `if (!parameters.ventilating)` branch from `Pufferfish::Driver::Power::transform`, everything still seems to behave correctly. I don't see any spamming of the event log; and I get this alarm on the quickstart screen, while ventilation is inactive. The only weird behavior I see here is that I get the "Charger is disconnected" event twice in the event log, occurring 23 seconds apart.

Results from testing on commit 352f3a3, with `ltc4015_sensor` removed from the list of firmware initializables:
- When I remove the `if (!parameters.ventilating)` branch from `Pufferfish::Driver::Power::transform` (and delete all code in `deactivate_alarms`, to make the commit compile), the event log is **not** spammed with "Charger is disconnected" events.

Results from testing on commit 2abfd33, with `ltc4015_sensor` removed from the list of firmware initializables:
1. When I run the firmware without any modifications, the firmware simulator for power management does not seem to work: the battery level is always shown as 100%, and charging is always `false`. I got a single alarm that the battery charger is disconnected, but the event log wasn't spammed by that alarm.
2. When I remove the `if (!parameters.ventilating)` branch from `Pufferfish::Driver::Power::transform`, the event log is spammed with "Charger is disconnected" events, but only when the ventilator is not ventilating; once the ventilator starts ventilating, it works fine
3. When I run the simulator backend, the power management simulation seems to work, and the alarms seem to work.

Results from testing on commit 01948ce, with `ltc4015_sensor` removed from the list of firmware initializables:
- This commit doesn't have a `if (!parameters.ventilating)` branch in `Pufferfish::Driver::Power::transform`. Battery level is always at 100%, and the event log is spammed with "Charger is disconnected" events.

Thus, some change made on 352f3a3 fixed the problem of spamming the event log even without the `if (!parameters.ventilating)` branch. My hypothesis is that `BreathingCircuit::AlarmsService::deactivate_alarms` was always deactivating the `LogEventCode::charger_disconnected` alarm whenever the breathing circuit found that the parameters had set `ventilating` to false. Thus, the changes made in 2abfd33 to remove `charger_disconnected` from the list of alarms which the Breathing Circuit alarms service could deactivate had a side effect of also solving the root cause of the event log spamming issue. By adding that event code back to the list of alarm codes for the Breathing Circuit alarm service, I was able to bring the spamming back. This also explains why your workaround to disable the `charger_disconnected` alarm was able to stop the spamming: then your Power alarms service was no longer fighting against the Breathing Circuit alarms service about whether to activate or deactivate `charger_disconnected`.

Since we have identified the root cause of the event log being spammed by `charger_disconnected` events, it is now correct to remove the `if (!parameters.ventilating)` branch from `Pufferfish::Driver::Power::transform`.

rohanpurohit committed:
Power: sound alarm when not ventilating as well

rohanpurohit committed:
Power: update simulator to fix discharging

rohanpurohit committed:
Sensor: cleanup LTC4015 high level driver

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-driver-power-management

rohanpurohit committed:
firmware: resolve clang-tidy issues

rohanpurohit committed:
firmware: apply formatting autofix

ethanjli commented on firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Power/Simulator.cpp:
```diff
@@ -0,0 +1,75 @@
+std::uniform_real_distribution<float> uniform_centered(-uniform_width / 2, uniform_width / 2);
+
+void Simulator::input_clock(uint32_t current_time) {
+  if (initial_time_ == 0) {
```
The `initial_time` related code in the simulator backend was only so that the simulator backend's timestamps would resemble the firmware's timestamps, as Python's `time.time()` has time=0 at Jan 1 1970 while the firmware's `HAL::time()` has time=0 when the firmware first started. We should remove `initial_time_` from the firmware's power simulator.

ethanjli replied:
Actually, the Breathing Circuit simulator also has `initial_time` (though maybe it shouldn't either). We can leave `initial_time` in the power simulator for this PR, just to be consistent.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Power/Simulator.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+
+void Simulator::transform(uint32_t current_time, PowerManagement &power_management) {
+  input_clock(current_time);
```
When you call `input_clock` here and it determines that an update is needed, then it sets `previous_time_ = current_time_`. Next, you call `transform_charge`/`transform_discharge`, which does nothing if `previous_time_ == current_time_` (as determined by an `update_needed()` check). I haven't tried running the latest version of the firmware on this PR yet, but this code seems suspicious to me.

ethanjli replied:
It looks like this same pattern is used in the BreathingCircuit's simulator, so that needs further investigation too.

ethanjli replied:
Looks like I missed a basic thing which explains all of this: right after `previous_time` is set to `current_time_`, `current_time_` is changed to a new value.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Power/Simulator.cpp:
```diff
@@ -0,0 +1,75 @@
+  }
+  power_management.charging = true;
+  power_management.power_left +=
+      (1 + (power_responsiveness * uniform_centered(prng))) / uniform_int(prng);
```
This formula is seems more convoluted than necessary. If you're simply trying to add a random percentage to `power_left`, it would be clearer to just define a uniform random distribution ` std::uniform_real_distribution<float> uniform_percentage(min_increase_per_step, max_increase_per_step)` (and you can just delete `uniform_int` and `uniform_centered`) and add `uniform_percentage(prng)` to `power_left`.

rohanpurohit replied:
I'm now simply incrementing from 0 to 100 every second and decrementing on discharge, works fine and we can get rid of random distributions and randomness in results too, I'm resolving this!

ethanjli commented on firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Power/Simulator.cpp:
```diff
@@ -0,0 +1,75 @@
+  power_management.charging = false;
+  power_management.power_left -=
+      (1 + (power_responsiveness * uniform_centered(prng))) / uniform_int(prng);
+  if (power_management.power_left == 0) {
```
If `transform_discharge` should be symmetric with `transform_charge`, should this `if` check be `power_management.power_left <= 0`? And should the body of the if branch also set `power_management.power_left = 0?

rohanpurohit replied:
I had done the same previously, although `cppcheck` complaints about `Core/Src/Pufferfish/Driver/Power/Simulator.cpp:53:35: style: Checking if unsigned expression 'power_management.power_left' is less than zero. [unsignedLessThanZero]
`

ethanjli replied:
Ah, ok - I had thought `power_management.power_left` was a `float` rather than a `uint*_t`. Do you think we should change it to `float`, or leave it as `uint*_t`?

rohanpurohit replied:
Let's change it to `float`?

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/HAL/Interfaces/I2CDevice.h:
```diff
@@ -43,6 +43,15 @@ class I2CDevice {
   virtual I2CDeviceStatus read(uint8_t *buf, size_t count) = 0;
 
+  /**
+   * Reads data from the device
```
It would be helpful to add a note here about when this `read` method should be used instead of the other version of the `read` declared on line 44.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Power/Simulator.h:
```diff
@@ -0,0 +1,36 @@
+class Simulator {
+ public:
+  void transform(uint32_t current_time, PowerManagement &power_management);
+  void transform_charge(PowerManagement &power_management);
+  void transform_discharge(PowerManagement &power_management);
```
These should be private methods.

ethanjli commented on proto/mcu_pb.proto:
```diff
@@ -186,9 +187,9 @@ message ActiveLogEvents {
 // Power Management
 
-message BatteryPower {
+message PowerManagement {
```
Maybe it would be better to name this `MCUPowerStatus` (in the future we'll also be adding a `GUIPowerStatus` message for the Raspberry Pi, which is powered by a separate battery and a separate copy of the LTC4015 chip).

ethanjli commented on firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/I2C/LTC4015/Device.cpp:
```diff
@@ -0,0 +1,102 @@
+  uint16_t vbat_value = 0;
+  Util::read_bigend(buffer.data(), vbat_value);
+  // NOLINTNEXTLINE(readability-magic-numbers)
+  v_bat = vbat_value * 128.176;  // Vbat = [VBAT] • 128.176µV for lead-acid
```
All magic numbers for this driver should be defined as static constants, either private in the class declaration of `Device` (if the numbers aren't needed elsewhere) or in `Types.h` (if the numbers are needed elsewhere).

ethanjli replied:
It would be great if you could add code comments to these formulas (and/or to the definitions of static constants for magic numbers) indicating where to find these formulas in the datasheet (e.g. a URL and page number)

rohanpurohit replied:
Yep! added link for datasheet along with page number, they are only required now in `Device` so i have added it there.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/I2C/LTC4015/Sensor.h:
```diff
@@ -0,0 +1,60 @@
+
+ private:
+  Action next_action_ = Action::initialize;
+  uint32_t current_time_us_ = 0;
```
Do you still need `current_time_us_` in `StateMachine`? It doesn't look like it's being used anywhere.

rohanpurohit replied:
Yep not using it, cleaned up the driver

ethanjli commented on firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/I2C/LTC4015/Sensor.cpp:
```diff
@@ -0,0 +1,88 @@
+
+InitializableState Sensor::output(PowerManagement &power_management) {
+  switch (next_action_) {
+    default:
```
Since you put the `default` case up here, I think you'll always fall through to `case Action::measure` and return from inside it, and you'll never reach the `return InitializableState::failed` line at the end of this method. Maybe you meant to move the `default` case to the end of the `switch` statement?

rohanpurohit replied:
Yes fixed!

ethanjli reviewed (commented):
Here's my review of the code. The firmware implementation for #322 depends on modifying the Alarms Manager (which was refactored in this PR), so let's discuss how we should proceed (e.g. should we make a separate PR with just the Alarms Manager refactor and merge that in immediately, or should I copy the Alarms Manager refactor onto my branch for #322, or should we just merge my branch for #322 after we merge in #327, or some other strategy)

ethanjli commented:
Results from testing the power simulator in the firmware: simulated charging happens slowly (I think it's a reasonable charging rate), while discharging happens way too quickly (i.e. it goes from 100% to 0% in a few seconds). This bug only occurs in the firmware's power simulator, not in the simulator backend.

rohanpurohit committed:
firmware: cleanup ltc4015 sensor

rohanpurohit committed:
trivial changes to power simulator

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-driver-power-management

rohanpurohit committed:
Device: add links and get rid of magic numbers

rohanpurohit committed:
Device.h: update consts to conxtexpr

rohanpurohit committed:
cleanup and update LTC4015 low/high drivers

rohanpurohit committed:
power: cleanup and fix power simulator

rohanpurohit commented:
@ethanjli I have updated the power simulator and higher-level driver, this is ready for testing and code review, I will update the changes you suggested in #378 once we test and merge that.

rohanpurohit committed:
resolve cppcheck complaint

rohanpurohit committed:
rename PowerManagement to MCUPowerStatus

rohanpurohit committed:
update configuration of power simulator in main.cpp

ethanjli commented on firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/I2C/LTC4015/Device.cpp:
```diff
@@ -0,0 +1,117 @@
+  v_bat =
+      vbat_value *
+      static_cast<uint16_t>(
+          batt_voltage_conversion);  // https://www.analog.com/media/en/technical-documentation/data-sheets/4015fb.pdf
```
The formatting here is weird, I think clang-format will fix it if you move the comment to its own line before the line where you assign `v_bat`. This also applies to the other comments in this file.

ethanjli reviewed (commented):
Here's my review, just minor formatting changes. I tested the backend simulator and firmware simulator and they both work fine now. If we merge this into develop ahead of #378, I'll resolve any merge conflicts and make any changes in #378. For records-keeping:

1. This project is licensed under Apache License v2.0 for any software, and Solderpad Hardware License v2.1 for any hardware - do you agree that your contributions to this project will be under these licenses, too?
2. Were any of these contributions also part of work you did for an employer or a client?
3. Does this work include, or is it based on, any third-party work which you did not create?

rohanpurohit committed:
Device: fix formatting issue

rohanpurohit committed:
change power left from uint32 to float

rohanpurohit commented:
> Here's my review, just minor formatting changes. I tested the backend simulator and firmware simulator and they both work fine now. If we merge this into develop ahead of #378, I'll resolve any merge conflicts and make any changes in #378. 

either way is fine with me, I guess we need to use `Util::TImer` in the simulator and remove `Util::UnrecognizedMessage` as now it uses EnumMap.

1. This project is licensed under Apache License v2.0 for any software, and Solderpad Hardware License v2.1 for any hardware - do you agree that your contributions to this project will be under these licenses, too? Yes.
2. Were any of these contributions also part of work you did for an employer or a client? No.
3. Does this work include, or is it based on, any third-party work which you did not create? No.

ethanjli commented:
I just merged #378 into develop before I saw your latest comment. One other change which was buried in that PR's description is that we should change `AlarmsService::transform(PowerManagement &power_management,...` to `AlarmsService::transform(const PowerManagement &power_management,...` if that alarms service only needs to read values from `power_management` and doesn't need to write values to it.

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-driver-power-management

rohanpurohit committed:
Power: alarms: pass const reference for MCUPowerStatus

rohanpurohit committed:
Simulator: use Util:Timer instead of within_timeout

rohanpurohit committed:
firmware: fix includes for make_array

rohanpurohit committed:
firmware: fix formatting

rohanpurohit committed:
main.cpp: trivial changes

rohanpurohit committed:
main.cpp: fix formatting

rohanpurohit committed:
main: resolve clang-tidy issue
